---
title: "Assignment 2"
output: html_document
date: '2022-09-21'
---
Install and attach all required libs using librarian package
```{r Install and attach project libraries}
# Script for installing required R packages for this project
install.packages("librarian")
install.packages("BiocManager") # For using librarian to install Bioconductor packages
BiocManager::install("Biobase") # For using librarian to install Bioconductor packages
librarian::shelf(DESeq2, apeglm, ggplot2, magrittr, tidyverse, rstudioapi, dplyr, tidyr, EnhancedVolcano, gprofiler2, topGO, org.Hs.eg.db)
```
Set seed for reproducibility
```{r Set seed}
set.seed(12345)
```

Get the CSVs [generated in Python](./Assignment_2.ipynb#export-to-r) into data frames
```{r Load data from CSVs}
# Set the path of the working directory
setwd(dirname(getActiveDocumentContext()$path))

# Read the csv files to dfs
data <- read_csv("../Data/data_event_sorted.csv") %>% tibble::column_to_rownames("index")
clinical <- read_csv("../Data/clinical_event_sorted.csv") %>% as.data.frame(clinical)
names(clinical)[names(clinical) == "Primary Cytogenetic Code"] <- "PCC" # Rename PCC column
```

Initializing DESeq object
```{r Initialize DESeq object}
ddset <- DESeqDataSetFromMatrix(
  countData = data,
  colData = clinical,
  design = ~PCC
)
```

Pre-filtering ddset for efficiency
```{r Pre-filtering}
ddset <- ddset[rowSums(counts(ddset) > 10, na.rm = TRUE) > 7, ] # Remove genes with fewer than 7 samples/10 counts
save(ddset, file = "deseq_set.RData") # Save the pre-filtered dataset to use in Assignment 3
```

Creating DESeq object and storing its results
```{r Make DESeq object, echo=False}
deseq_object <- DESeq(ddset)
deseq_results <- results(deseq_object)
save(deseq_object, file = "deseq_object.RData") # Store object for efficiency
```

Here we will use `lfcShrink()` function to obtain shrunken log fold change estimates based on negative binomial distribution.
This will add the estimates to your results table. 
Using `lfcShrink()` can help decrease noise and preserve large differences between groups (it requires that `apeglm` package be installed) [@Zhu2018].
```{r Obtaining LFC}
deseq_results <- lfcShrink(
  deseq_object, # The original DESeq2 object after running DESeq()
  coef = 6, # The log fold change coefficient used in DESeq(); the default is 2.
  res = deseq_results # The original DESeq2 results table
)
```

```{r}
summary(deseq_results)
```

```{r DESeq2 results}
head(deseq_results)
```


Cleaning the results to suit what we want
```{r Cleaning results}
# this is of class DESeqResults -- we want a data frame
deseq_df <- deseq_results %>%
  # make into data.frame
  as.data.frame() %>%
  # the gene names are row names -- let's make them a column for easy display
  tibble::rownames_to_column("Gene") %>%
  # add a column for significance threshold results
  dplyr::mutate(threshold = padj < 0.05) %>%
  # sort by statistic -- the highest values will be genes with
  # higher expression in RPL10 mutated samples
  dplyr::arrange(dplyr::desc(log2FoldChange))
```

```{r Viewing cleaned results}
head(deseq_df)
```

Gene expressions associated with different types of AML
```{r Getting gene expressions}
gene_expression <- function(gene_name) {
  file <- paste("../Figures/Gene/", gene_name, ".png")
  png(file)
  DeSeq2::plotCounts(ddset, gene = gene_name, intgroup = "PCC")
  dev.off()
}

for (x in row.names(df_by_var))
  gene_expression(x)
```


Volcano Plot
```{r Generating volcano plot}
#png("../Plots/pretty_volcano_plot.png")
volcano_plot <- EnhancedVolcano::EnhancedVolcano(
  deseq_df,
  lab = deseq_df$Gene,
  x = "log2FoldChange",
  y = "padj",
  pCutoff = 0.01,
  pointSize = 3.0,
  legendLabels=c('Not sig.','Log2FC','p-value',
      'p-value & Log2FC'),
  title = "Differential expression",
  subtitle = bquote(italic("AML primary cytogenetic code volcano plot")),
  colAlpha = 4/5,
)
```

```{r}
volcano_plot
```

GENE ENHANCEMENT -- Choosing 20 most variant genes ([And plot the heatmap in Python](./Assignment_2.ipynb#heat-map))
```{r Extracting differentially expressed genes}
# Normalizing data
dds_norm <- vst(ddset)

# Calculate the variance for each gene
variances <- apply(assay(dds_norm), 1, var)

# Determine the upper quartile variance cutoff value
upper_var <- quantile(variances, 0.99943)

# Filter the data choosing only genes whose variances are in the upper quartile
df_by_var <- data.frame(assay(dds_norm)) %>%
  dplyr::filter(variances > upper_var)
```

```{r Saving the list of DEGs}
setwd(dirname(getActiveDocumentContext()$path))
write.csv(df_by_var, '../Data/significant_genes.csv', row.names = TRUE)
```


- GENE ENHANCEMENT BY GPROFILER2
```{r Gene Enhancement with gprofiler2}
gostres <- gost(row.names(data),
  organism = "hsapiens",
  multi_query = FALSE, significant = TRUE, exclude_iea = FALSE,
  measure_underrepresentation = FALSE, evcodes = FALSE,
  user_threshold = 0.05, correction_method = "g_SCS",
  domain_scope = "annotated", custom_bg = NULL,
  numeric_ns = "", sources = NULL, as_short_link = FALSE
)
```

```{r}
names(gostres)
head(gostres$result, 20)
row.names(data)
```

GENE ENHANCEMENT ANALYSIS BY topGO
Load the deseq_df to avoid generating new deseq objects
```{r Initializing DEGs for enhancement using topGO}
load("df.RData")
deseq_df <- na.omit(deseq_df)
```
Generate factor based on p-vals and gene names
```{r Generating p-val and gene name factor}
gene_list <- as.integer(deseq_df$padj)
names(gene_list) <- deseq_df$Gene
```
Function for selecting significant genes
```{r Select significant genes}
topDiffGenes <- function(allScore) {
  return(allScore < 0.01)
}
```
Create GO object 
```{r Creating topGO object}
godata <- new("topGOdata",
  ontology = "MF",
  allGenes = gene_list,
  geneSel = topDiffGenes,
  annot = annFUN.org,
  mapping = "org.Hs.eg.db",
  ID = "symbol"
)
```
Extract list of significant genes
```{r Extracting significant DEGs}
sig_genes <- sigGenes(godata)
str(sig_genes)
numSigGenes(godata)
```

