---
title: "A_2"
output: html_document
date: '2022-09-21'
---

Attaching all required libraries and setting seed for reproducibility
```{r attach-libs}
# run install.packages('BiocManager) && BiocManager::install("DESeq2") and ("apeglm") ("EnhancedVolcano")
# Attach the DESeq2 library
library(DESeq2)

# Attach the ggplot2 library for plotting
library(ggplot2)

# We will need this so we can use the pipe: %>%
library(magrittr)

# Tidyverse
library(tidyverse)

# For automatically setting the working directory path
library(rstudioapi)

set.seed(12345)
```

Get the CSVs [generated in Python](./Assignment_2.ipynb#export-to-r) into data frames
```{r get-data}
# Set the path of the working directory
setwd(dirname(getActiveDocumentContext()$path))

# Read the csv files to dfs
data <- read_csv("../Data/data_event_sorted.csv") %>% tibble::column_to_rownames("index")
clinical <- read_csv("../Data/clinical_event_sorted.csv") %>% as.data.frame(clinical)
```

```{r rename-PCC}
names(clinical)[names(clinical) == "Primary Cytogenetic Code"] <- "PCC"
```

Initializing DESeq object
```{r init-deseq}
ddset <- DESeqDataSetFromMatrix(
  countData = data,
  colData = clinical,
  design = ~PCC
)
```

Creating DESeq object and storing its results
```{r make-object, echo=False}
deseq_object <- DESeq(ddset)
deseq_results <- results(deseq_object)
```

Here we will use `lfcShrink()` function to obtain shrunken log fold change estimates based on negative binomial distribution.
This will add the estimates to your results table. 
Using `lfcShrink()` can help decrease noise and preserve large differences between groups (it requires that `apeglm` package be installed) [@Zhu2018].
```{r obtain-lfc}
deseq_results <- lfcShrink(
  deseq_object, # The original DESeq2 object after running DESeq()
  coef = 6, # The log fold change coefficient used in DESeq(); the default is 2.
  res = deseq_results # The original DESeq2 results table
)
```

```{r view-results}
head(deseq_results)
```


Cleaning the results to suit what we want
```{r clean-results}
# this is of class DESeqResults -- we want a data frame
deseq_df <- deseq_results %>%
  # make into data.frame
  as.data.frame() %>%
  # the gene names are row names -- let's make them a column for easy display
  tibble::rownames_to_column("Gene") %>%
  # add a column for significance threshold results
  dplyr::mutate(threshold = padj < 0.05) %>%
  # sort by statistic -- the highest values will be genes with
  # higher expression in RPL10 mutated samples
  dplyr::arrange(dplyr::desc(log2FoldChange))
```

```{r view-cleaned-results}
head(deseq_df)
```

Gene expressions associated with different types of AML
```{r get-associated-expressions}
gene_expression <- function(gene_name) {
  file <- paste("../Figures/Gene/", gene_name, ".png")
  png(file)
  plotCounts(ddset, gene = gene_name, intgroup = "PCC")
  dev.off()
}

for (x in row.names(df_by_var))
  gene_expression(x)
```


Volcano Plot
```{r volcano-plot}
#png("../Plots/pretty_volcano_plot.png")
volcano_plot <- EnhancedVolcano::EnhancedVolcano(
  deseq_df,
  lab = deseq_df$Gene,
  x = "log2FoldChange",
  y = "padj",
  pCutoff = 0.01,
  pointSize = 3.0,
  legendLabels=c('Not sig.','Log2FC','p-value',
      'p-value & Log2FC'),
  title = "Differential expression",
  subtitle = bquote(italic("AML primary cytogenetic code volcano plot")),
  colAlpha = 4/5,
)
```

```{r view-volcano-plot}
volcano_plot
```

GENE ENHANCEMENT -- Choosing 20 most variant genes ([And plot the heatmap in Python](./Assignment_2.ipynb#heat-map))
```{r extract-diff-expression}
# Normalizing data
dds_norm <- vst(ddset)

# Calculate the variance for each gene
variances <- apply(assay(dds_norm), 1, var)

# Determine the upper quartile variance cutoff value
upper_var <- quantile(variances, 0.99943)

# Filter the data choosing only genes whose variances are in the upper quartile
df_by_var <- data.frame(assay(dds_norm)) %>%
  dplyr::filter(variances > upper_var)
```

```{r save-list}
setwd(dirname(getActiveDocumentContext()$path))
write.csv(df_by_var, '../Data/significant_genes.csv', row.names = TRUE)
```


- GENE ENHANCEMENT BY GPROFILER2

  - Install Package
```{r attach-lib-for-enhancement}
# install.packages("gprofiler2")
library(gprofiler2)
```


```{r fernando-enhancement}
gostres <- gost(row.names(data),
  organism = "hsapiens",
  multi_query = FALSE, significant = TRUE, exclude_iea = FALSE,
  measure_underrepresentation = FALSE, evcodes = FALSE,
  user_threshold = 0.05, correction_method = "g_SCS",
  domain_scope = "annotated", custom_bg = NULL,
  numeric_ns = "", sources = NULL, as_short_link = FALSE
)
```

```{r view-fernando-enhancement}
names(gostres)
head(gostres$result, 20)
row.names(data)
```

GENE ENHANCEMENT ANALYSIS BY topGO
```{r attach-libs-for-topGO-enhancement}
BiocManager::install("topGO")
BiocManager::install("org.Hs.eg.db")
library(topGO)
library(org.Hs.eg.db)
```
Load the deseq_df to avoid generating new deseq objects
```{r}
load("df.RData")
deseq_df <- na.omit(deseq_df)
```
Generate factor based on p-vals and gene names
```{r}
gene_list <- as.integer(deseq_df$padj)
names(gene_list) <- deseq_df$Gene
```
Function for selecting significant genes
```{r}
topDiffGenes <- function(allScore) {
  return(allScore < 0.01)
}
```
Create GO object 
```{r}
godata <- new("topGOdata",
  ontology = "MF",
  allGenes = gene_list,
  geneSel = topDiffGenes,
  annot = annFUN.org,
  mapping = "org.Hs.eg.db",
  ID = "symbol"
)
```
Extract list of significant genes
```{r}
sig_genes <- sigGenes(godata)
str(sig_genes)
numSigGenes(godata)
```
Graph godata
```{r}
graph(godata)
```
